### PR TITLE
[FLINK-29721][hive] Supports native hive min function for hive dialect

### PR DESCRIPTION
**Pull request by Tartarus0zm**
## What is the purpose of the change

Supports native hive min function for hive dialect. Due to flink doesn't support compare array and row type, so this implementation also doesn't min array and row type, but hive udaf supports it.

## Brief change log

* Supports native hive min function for hive dialect

## Verifying this change

This change added tests and can be verified as follows:

* Added integration tests in HiveDialectQueryITCase

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by Tartarus0zm** [comment]
@lsyldliu @beyond1920  hello, please take a look, if you have time, thanks

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/module/hive/HiveModule.java [hunk 1]
```diff
@@ -81,6 +82,18 @@ public class HiveModule implements Module {
                                     "tumble_rowtime",
                                     "tumble_start")));
 
+    static final Set<String> FLINK_BUILT_IN_FUNC_FOR_HIVE =
```
Comment (lsyldliu) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/module/hive/HiveModule.java:85]: In [FLINK-30579](https://github.com/apache/flink/pull/21629) pr, we will introduce an option to enable native hive sum aggregate function, so I think this set should only contain the agg function name, and exclude other functions.

**Review by lsyldliu — COMMENTED** [review]
@Tartarus0zm Thanks for the contribution, I left one comment.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/module/hive/HiveModule.java (hunk 1 above)
Comment (Tartarus0zm) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/module/hive/HiveModule.java:85]: I've address your comments.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/module/hive/HiveModule.java [hunk 2]
```diff
@@ -81,6 +82,9 @@ public class HiveModule implements Module {
...
```
Comment (lsyldliu) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/module/hive/HiveModule.java:85]:
```suggestion
    static final Set<String> BUILTIN_NATIVE_AGG_FUNC =
```

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/module/hive/HiveModule.java [hunk 3]
```diff
@@ -178,4 +182,19 @@ public Optional<FunctionDefinition> getFunctionDefinition(String name) {
     public String getHiveVersion() {
         return hiveVersion;
     }
+
+    private Optional<FunctionDefinition> getFlinkBuiltInFunction(
```
Comment (lsyldliu) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/module/hive/HiveModule.java:186]:
```suggestion
    private Optional<FunctionDefinition> getBuiltInNativeAggFunction(
```

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/module/hive/HiveModule.java [hunk 4]
```diff
@@ -178,4 +182,19 @@ public Optional<FunctionDefinition> getFunctionDefinition(String name) {
...
+        switch (name) {
+            case "sum":
+                // We override Hive's sum function by native implementation to supports hash-agg
+                return Optional.of(new HiveSumAggFunction());
+            case "min":
+                // We override Hive's min function by native implementation to supports hash-agg
+                return Optional.of(new HiveMinAggFunction());
+            default:
+                throw new UnsupportedOperationException(
+                        "flink built-in hive function not support " + name + " yet!");
```
Comment (lsyldliu) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/module/hive/HiveModule.java:197]:
```suggestion
                        String.format("Built-in hive aggregate function doesn't support %s yet!", name);
```

**Comment by Tartarus0zm** [comment]
@flinkbot run azure

**Review by lsyldliu — APPROVED** [review]
LGTM

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectQueryITCase.java [hunk 5]
```diff
@@ -1062,6 +1062,87 @@ public void testSumAggWithGroupKey() throws Exception {
         tableEnv.executeSql("drop table test_sum_group");
     }
 
+    @Test
+    public void testMinAggFunctionPlan() {
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectQueryITCase.java:1066]: it's better the plan test should not be in ITCase

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectQueryITCase.java [hunk 6]
```diff
@@ -1062,6 +1062,87 @@ public void testSumAggWithGroupKey() throws Exception {
...
+        // test explain
+        String actualPlan = explainSql("select x, min(y) from foo group by x");
+        assertThat(actualPlan).isEqualTo(readFromResource("/explain/testMinAggFunctionPlan.out"));
+    }
+
+    @Test
+    public void testMinAggFunction() throws Exception {
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectQueryITCase.java:1073]: all type should be cover, including unsupported types

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveMinAggFunction.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+    @Override
+    public Expression[] retractExpressions() {
+        throw new TableException("Min aggregate function does not support retraction.");
+    }
+
+    @Override
+    public Expression[] mergeExpressions() {
+        return new Expression[] {
+            /* max = */ ifThenElse(
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveMinAggFunction.java:90]: min

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribution, I left some minor comments

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectQueryITCase.java (hunk 5 above)
Comment (lsyldliu) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectQueryITCase.java:1066]: Current Hive ITCase covered all the plan related tests, we can improve this by another separate PR

**Review by godfreyhe — APPROVED** [review]
Thanks for the update, LGTM

**Comment by Tartarus0zm** [comment]
@godfreyhe  I've rebase the code, please take a look again.

**Pull request closed without merging**